Expensify/App
# Refactor date picker to be compatible with form

### Details

> 1. An optional isFormInput prop.
> 2. If isFormInput is true, require a inputID prop. Use [getInputIDPropTypes] (https://github.com/Expensify/App/blob/1860b590114991d1819a5590608d26c32337de56/src/libs/FormUtils.js#L7)[] to enforce this propType rule.
>3. Add an optional shouldSaveDraft prop. Defaults to false.

Importing the `baseTextInputPropTypes` does the deal. I have added a `propType` for defaultValue to support both dates and strings as consistency in the value and defaultValue parameter is expected.

>4. Make the value prop optional.

The `value` prop is not a required prop.

>5.In the input’s onBlur method, call props.onBlur().
>6.In the input’s onChange method (or equivalent e.g. onTextChange, etc), call props.onChange().
>7. Remove the hasError prop.

Done!
>8. Add an optional errorText prop. Defaults to an empty string.

Importing the `baseTextInputPropTypes` does the deal. 

> 9.Update the error message to display if errorText is truth.
Error text is displayed only when non-empty!

>10.  Update the inline error display style so it is left aligned with the label and input value. 

Done! Strangely it hadn't been implemented for TextInput in the PR.

>11. Make sure that props.ref is attached to the appropriate DOM node. This could involve forwarding the ref to a child component. This is important so we can call ref.focus().

Done! Ref has been passed to TextInput elements because DatePicker doesn't natively support focus. These are handled via different methods. t`his.textInput.clear() `and similar methods are accessible

> 12. Add the input to the test [Form stories](https://github.com/Expensify/App/blob/1860b590114991d1819a5590608d26c32337de56/src/stories/Form.stories.js) and make sure that it works. You can check this by running npm run storybook and testing the component in the example form.

Done!
>13 Remove any unused code.

Done!

There are a few more discretionary changes done. I will mention those in the comments.

**Newly discovered Issues:** 
1.The `datepicker` doesn't get focussed on clicking in android/iOS. 
2. The storybook sidebar UI styling is broken for mWeb.

### Fixed Issues

$ https://github.com/Expensify/App/issues/7536

### Tests


1. Run npm run storybook
2.  Play around with the Form stories and make sure that it works as intended. Specifically, note the following:
- We can enter values into the inputs.
- Clicking submit displays a loading indicator in the submit button and an alert with the correct input values.
- Emptying the input value and blurring it shows an error.
- Clicking the fix the errors link above the submit button focus on the first input with an error.
- Entering a value in the input clears the error.
- Verify that server errors are displayed correctly.
- Any other fishy behavior.

Run the app and test that DatePicker still works fine.

The errors observed were common to `TextInput`. Resolution of those should rectify the ones that appear here.
### QA Steps



### Tested On


### Screenshots


#### Web


https://user-images.githubusercontent.com/48352577/152928931-7cb47f9a-6be1-4f5b-a2c9-3c688790de98.mp4


#### Mobile Web


https://user-images.githubusercontent.com/48352577/152930095-7b7444a0-2a4c-4a15-9b36-9f6b29247bef.mp4


#### Desktop

https://user-images.githubusercontent.com/48352577/152926818-62d253b8-f7cd-42ca-a6bd-fbad8c4c3a60.mp4
#### iOS

https://user-images.githubusercontent.com/48352577/152926171-b99a7c18-9db4-40d9-b497-981499ebc5c7.mp4
#### Android


https://user-images.githubusercontent.com/48352577/152925707-df51003c-9539-4a45-9eb8-add5c252be9f.mp4
